RecoEgamma/EgammaTools/src/HGCalIsoProducer.cc

```diff
@@ -10,51 +10,83 @@
         throw std::runtime_error("HGCalIsoProducer::produceHGCalIso: rechittools not set");
+    
+    hitEtaPhiCache_.clear();
+    hitEtaPhiCache_.reserve(recHitsEE_->size()+recHitsFH_->size()+recHitsBH_->size());
+
+    // Since HGCal is not projective and the rechits don't cache any
+    // eta,phi, we make our own here
+    auto makeEtaPhiPair = [this](const auto& hit) {
+        const GlobalPoint position = rechittools_->getPosition(hit.id());
+        float eta=rechittools_->getEta(position, 0); //assume vertex at z=0
+        float phi=rechittools_->getPhi(position);
+        return std::make_pair(eta, phi);
+    };
 
-    for(auto& r:isoringdeposits_)
-        r=0;
-
+    for(const auto& hit : *recHitsEE_) hitEtaPhiCache_.push_back(makeEtaPhiPair(hit));
+    for(const auto& hit : *recHitsFH_) hitEtaPhiCache_.push_back(makeEtaPhiPair(hit));
+    for(const auto& hit : *recHitsBH_) hitEtaPhiCache_.push_back(makeEtaPhiPair(hit));
+}
 
-    //this could be replaced by the hit map created by storeRecHits in PCAhelpers
-    std::vector<DetId>seedhits;
-    const std::vector<std::pair<DetId,float > > & seedhitmap= seed->hitsAndFractions();
-    for(const auto& h:seedhitmap)
-        seedhits.push_back(h.first);
+void HGCalIsoProducer::produceHGCalIso(const reco::CaloClusterPtr & seed){
 
-    for(const auto& hit: *allHitMap_) {
+    if(!rechittools_)
+        throw std::runtime_error("HGCalIsoProducer::produceHGCalIso: rechittools not set");
 
-        const GlobalPoint position = rechittools_->getPosition(hit.first);
-        float eta=rechittools_->getEta(position, 0);//assume vertex at z=0
-        float phi=rechittools_->getPhi(position);
+    for(auto& r:isoringdeposits_)
+        r=0;
 
-        float deltar2=reco::deltaR2(eta,phi,seed->eta(),seed->phi());
+    // compiler couldn't figure out seed is const and
+    // recomputed log(atan2()) every time...
+    const float seedEta=seed->eta();
+    const float seedPhi=seed->phi();
+    const std::vector<std::pair<DetId, float>> & seedhitmap= seed->hitsAndFractions();
 
-        if(deltar2>dr2_ || deltar2<mindr2_) continue;
+    auto checkAndFill = [this, &seedEta, &seedPhi, &seedhitmap](const HGCRecHit& hit, std::pair<float,float> etaphiVal) {
+        float deltar2=reco::deltaR2(etaphiVal.first,etaphiVal.second,seedEta,seedPhi);
+        if(deltar2>dr2_ || deltar2<mindr2_) return;
 
-        size_t layer=rechittools_->getLayerWithOffset(hit.first);
-        if(layer>=nlayers_) continue;
+        size_t layer=rechittools_->getLayerWithOffset(hit.id());
+        if(layer>=nlayers_) return;
 
         const size_t& ring=ringasso_.at(layer);
 
         //do not consider hits associated to the photon cluster
-        if(std::find(seedhits.begin(),seedhits.end(),hit.first)==seedhits.end()){
-            isoringdeposits_.at(ring)+=hit.second->energy();
+        if( std::none_of(seedhitmap.begin(), seedhitmap.end(), [&hit](const auto& seedhit){return hit.id()==seedhit.first;}) ){
```

Comment:
@nsmith- You should use an std::unordered_map here, std::none_of can be very slow in certain cases. Using a hash map will solve most of your problems with HGCal hit-information lookups inside of loops.

Reply:
Hi @lgray,
Thanks for the suggestion.  Indeed `none_of` is the largest child of `produceHGCalIso`, but it is only 4% of it, with the loop itself dominating, if I am interpreting igprof correctly.  So I figured it didn't matter anymore.  If you think the sampling might be missing time in `none_of`, I'll give the map a try.
The reason I moved away from the map was to avoid a copy, which was big in the profile.  Perhaps an unordered map wouldn't be so expensive to construct since one can reserve.
```
    2             0.0  .........       0.05 / 0.12         PhotonIDHelper::computeHGCAL(reco::Photon const&, float) [1645]
    1             0.4  .........       1.01 / 1.28         ElectronIDHelper::computeHGCAL(reco::GsfElectron const&, float) [478]
8342  [525]       0.4       1.06       1.02 / 0.04       HGCalIsoProducer::produceHGCalIso(edm::Ptr<reco::CaloCluster> const&)
    1             0.0  .........       0.04 / 0.04         _ZSt9__find_ifIN9__gnu_cxx17__normal_iteratorIPKSt4pairI5DetIdfESt6vectorIS4_SaIS4_EEEENS0_5__ops10_Iter_predIZZN16HGCalIsoProducer15produceHGC      alIsoERKN3edm3PtrIN4reco11CaloClusterEEEENKUlRK9HGCRecHitS2_IffEE_clESN_SO_EUlRKT_E_EEESQ_SQ_SQ_T0_St26random_access_iterator_tag.isra.57 [2780]
    2             0.0  .........       0.01 / 0.01         hgcal::RecHitTools::getLayerWithOffset(DetId const&) const [5758]
```

Reply:
@nsmith- OK, yeah not a big issue for now, but to protect it from having tails or the sort with weird events it's probably best to collapse the computational complexity of the loop as much as possible. Not high priority, you should consider it for a follow up pull request.